relay-server/src/metrics_extraction/metrics_summary.rs

```diff
@@ -0,0 +1,511 @@
+        MetricsSummaryBucketValue {
+            min: summary
+                .min
+                .value()
+                .and_then(|min| FiniteF64::try_from(*min).ok()),
+            max: summary
+                .max
+                .value()
+                .and_then(|max| FiniteF64::try_from(*max).ok()),
+            sum: summary
+                .sum
+                .value()
+                .and_then(|sum| FiniteF64::try_from(*sum).ok()),
+            count: summary.count.value().cloned().unwrap_or(0),
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`BucketValue`] by aggregating the data into
+    /// a gauge like summary.
+    fn from_bucket_value(value: &BucketValue) -> MetricsSummaryBucketValue {
+        match value {
+            BucketValue::Counter(counter) => Self::from_counter(counter),
+            BucketValue::Distribution(distribution) => Self::from_distribution(distribution),
+            BucketValue::Set(set) => Self::from_set(set),
+            BucketValue::Gauge(gauge) => Self::from_gauge(gauge),
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`CounterType`].
+    fn from_counter(counter: &CounterType) -> MetricsSummaryBucketValue {
+        MetricsSummaryBucketValue {
+            min: Some(*counter),
+            max: Some(*counter),
+            sum: Some(*counter),
+            count: 1,
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`DistributionValue`].
+    fn from_distribution(distribution: &DistributionValue) -> MetricsSummaryBucketValue {
+        let mut min = FiniteF64::MAX;
+        let mut max = FiniteF64::MIN;
+        let mut sum = FiniteF64::new(0.0).unwrap();
+
+        for value in distribution {
+            min = std::cmp::min(min, *value);
+            max = std::cmp::max(max, *value);
+            sum = sum.saturating_add(*value);
+        }
+
+        MetricsSummaryBucketValue {
+            min: Some(min),
+            max: Some(max),
+            sum: Some(sum),
+            count: distribution.len() as u64,
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`SetValue`].
+    fn from_set(set: &SetValue) -> MetricsSummaryBucketValue {
+        // For sets, we limit to counting the number of occurrences.
+        MetricsSummaryBucketValue {
+            min: None,
+            max: None,
+            sum: None,
+            count: set.len() as u64,
+        }
+    }
+
+    /// Builds a [`MetricsSummaryBucketValue`] from a [`GaugeValue`].
+    fn from_gauge(gauge: &GaugeValue) -> MetricsSummaryBucketValue {
+        MetricsSummaryBucketValue {
+            min: Some(gauge.min),
+            max: Some(gauge.max),
+            sum: Some(gauge.sum),
+            count: gauge.count,
+        }
+    }
+
+    /// Merges two [`MetricsSummaryBucketValue`]s together by mutating `self` and consuming
+    /// `other`.
+    fn merge(&mut self, other: MetricsSummaryBucketValue) {
```

Comment:
I'd instead implement `std::ops::Add` and make the value `Copy`, immutability ftw.